packages/muelu/src/Transfers/Smoothed-Aggregation/MueLu_NullspaceFactory_def.hpp

```diff
@@ -114,6 +126,30 @@ namespace MueLu {
     const ParameterList & pL = GetParameterList();
     std::string nspName = pL.get<std::string>("Fine level nullspace");
 
+    // get coordinates and compute mean of coordinates. (or centroid).
+    
+    typedef typename Teuchos::ScalarTraits<Scalar>::coordinateType coordinate_type;
+    typedef Xpetra::MultiVector<coordinate_type,LO,GO,NO> RealValuedMultiVector;
+    RCP<RealValuedMultiVector> Coords;
+    ArrayRCP<const coordinate_type> xvals, yvals, zvals;
+    Scalar cx, cy, cz;
+
+    cx = 0.; cy = 0.; cz = 0.;
+    if(calculateRotations_) {
+      Coords = Get< RCP<RealValuedMultiVector> >(currentLevel, "Coordinates");
+
+      xvals  = Coords->getData(0);
+      cx     = Coords->getVector(0)->meanValue();
+      if (Coords->getNumVectors() > 1) {
+        yvals = Coords->getData(1);
+        cy    = Coords->getVector(1)->meanValue();
+      }
+      if (Coords->getNumVectors() > 2) {
+        zvals = Coords->getData(2);
+        cz    = Coords->getVector(1)->meanValue();
```

Comment:
Index `1` refers to y-direction, so maybe this is correct for the z-direction:
```suggestion
        cz    = Coords->getVector(2)->meanValue();
```

Reply:
Hmmm ... that does seem messed up. It's funny how my own personal tests passed. I actually did one with the ice sheet code that compares with the null space computed by the application with the one using muelu. Here, I got identical residual histories ... but this test was only 2D so it didn't test the 3rd dimension. The second test was with the MueLu_Driver comparing again the null space computed by MueLu and the one computed by Galeri. Again, this test gives identical residual histories. There is a chance that the centroid calculation only improves the conditioning of the basis ... but not the actual residual histories if numerical conditioning of the basis is not an issue.

Reply:
The best test is to run the code twice .... with Galeri rotations and MueLu rotations ... and see that the residuals are identical. I wasn't sure if this was easy to do with the Driver? Do you know?

Reply:
I don't know how to do this in the driver. However, by comparing the residual histories, you do not directly test the nullspace, only indirecltly.

For directly testing the correctness of the nullspace, you would need to add a unit test, e.g. in `packages/muelu/test/unit_tests/NullspaceFactory.cpp`. Unfortunately, there isn't a single unit test for the `NullspaceFactory` so far, so you would need to create the `packages/muelu/test/unit_tests/NullspaceFactory.cpp` file and add your test (and maybe some other tests as well). There, you could compare the generated nullspace to a hand-coded nullspace, that you know to be correct.